Enhance bluetooth host keys

ahmedmoheb-nordic opened the pull request:
Enhance the module
- Add API docs
- Fix minor issues while searching for a key
- Add assertions to check functions arguments

ahmedmoheb-nordic committed:
bluetooth: host: keys: Add APIs docs

Add API documentation to describe behavior, args and return value.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
bluetooth: host: keys: Add assertions to check arguments

Check arguments passed to APIs with assertions.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
bluetooth: host: keys: Check if returned key matches keys pool item

bt_keys_find_addr() is used to find key by both ID and address.
Following checks must continue to compare ID and address as well.
Or, we can compare key references which is faster.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
bluetooth: host: keys: Remove CONFIG_BT_MAX_CONN from key in use check

Eliminate key in use check dependency on (CONFIG_BT_MAX_CONN > 1).

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

alwa-nordic commented:
`bt_keys_find_irk` in 6ceacde2fc2c3a3798572274439544a2fdcb72f9 looks broken to me. It no longer resolves RPAs. In fact it's now equivalent to `bt_keys_find_addr` with an extra `if (!bt_addr_le_is_rpa(addr)) { return NULL; }`.

ahmedmoheb-nordic commented:
> `bt_keys_find_irk` in [6ceacde](https://github.com/zephyrproject-rtos/zephyr/commit/6ceacde2fc2c3a3798572274439544a2fdcb72f9) looks broken to me. It no longer resolves RPAs.

I think the older logic is still preserved 
- Firstly check if the address is an RPA
- Try to find a record with matched ID and address (that's the major change because the older one uses only the ID as a unique key, but the current implementation allows multiple items with the same ID if they differ in address)
- Then if  a matched record is found, check if the key type is an IRK key
- The rest is the same

Could you clarify your point? why do you think that the logic isn't preserved?

alwa-nordic commented:
> Could you clarify your point? why do you think that the logic isn't preserved?

- Assume `bt_addr_le_is_rpa(addr)` is true.
- Assume all bonds are resolved, i.e., `!bt_addr_le_is_rpa(key_pool[i].addr)` is true for all `i`.
- The above conditions imply `addr != key_pool[i].addr`  for all `i`.
- This implies `bt_keys_find_addr(id, addr)` will be `NULL` for all `id` and `addr`.
- This makes `bt_keys_find_irk` return NULL.
- Before the code change, `bt_keys_find_irk` would resolve `addr` and find the bond/bt_keys if it exists.

ahmedmoheb-nordic commented:
> > Could you clarify your point? why do you think that the logic isn't preserved?
> 
> * Assume `bt_addr_le_is_rpa(addr)` is true.
> * Assume all bonds are have the identity information, i.e., `!bt_addr_le_is_rpa(key_pool[i].addr)` is true for all `i`.
> * The above conditions imply `addr != key_pool[i].addr`  for all `i`.
> * This implies `bt_keys_find_addr(id, addr)` will be `NULL` for all `id` and `addr`.
> * This makes `bt_keys_find_irk` return NULL.
> * Before the code change, `bt_keys_find_irk` would resolve `addr` and find the bond/bt_keys if it exists.

As per my understanding based on your comment, the passed argument `addr` may not exist in the keys pool `key_pool[i].addr`. If that is true and intended, yes, you are right and the new implementation is for sure incorrect.

ahmedmoheb-nordic commented:
> > Could you clarify your point? why do you think that the logic isn't preserved?
> 
> * Assume `bt_addr_le_is_rpa(addr)` is true.
> * Assume all bonds are have the identity information, i.e., `!bt_addr_le_is_rpa(key_pool[i].addr)` is true for all `i`.
> * The above conditions imply `addr != key_pool[i].addr`  for all `i`.
> * This implies `bt_keys_find_addr(id, addr)` will be `NULL` for all `id` and `addr`.
> * This makes `bt_keys_find_irk` return NULL.
> * Before the code change, `bt_keys_find_irk` would resolve `addr` and find the bond/bt_keys if it exists.

Could you briefly describe the function `bt_keys_find_irk()` and its expected behavior so that I can update the documentation of it and also the unit testing test cases?

ahmedmoheb-nordic commented:
> `bt_keys_find_irk` in [6ceacde](https://github.com/zephyrproject-rtos/zephyr/commit/6ceacde2fc2c3a3798572274439544a2fdcb72f9) looks broken to me. It no longer resolves RPAs.

I've dropped that commit and updated the API docs, please check.

alwa-nordic commented:
> Could you briefly describe the function `bt_keys_find_irk()` and its expected behavior so that I can update the documentation of it and also the unit testing test cases?

The best description I can come up with from looking at the implementation follows.

`bt_keys_find_irk()`: Find bond by local identity and Peer RPA. \\ This finds a bond with the local identity `id` and Peer IRK that generated Peer RPA `addr`. This does not care about the `BT_KEYS_ID_ADDED` status. This guarantees not finding anything by identity address.

jori-nordic commented on subsys/bluetooth/host/keys.h:
```diff
@@ -6,6 +6,9 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#ifndef __BT_HOST_KEYS_H
```
we don't use include guards in zephyr

jhedberg replied:
We do, but generally not for internal header files, to catch circular dependencies more easily.

jhedberg commented on subsys/bluetooth/host/keys.h:
```diff
@@ -74,19 +77,108 @@ struct bt_keys {
+ * @param data User data to be passed to the callback function.
+ */
+void bt_keys_foreach_type(int type, void (*func)(struct bt_keys *keys, void *data),
 		     void *data);
```
Looks like the indentation/alignment of the split line is now wrong, after the function name got longer

carlescufi commented on subsys/bluetooth/host/keys.h:
```diff
@@ -74,19 +77,108 @@ struct bt_keys {
 
-void bt_keys_foreach(int type, void (*func)(struct bt_keys *keys, void *data),
+/**
+ * @brief Get a call through the callback for each key with the same type
```
Note that, in our case, it will not be generated even if we set `INTERNAL_DOCS` to `YES` because we only provide `include/` as input for Doxygen, but it's better to be ready in case this changes.

carlescufi replied:
@gmarull can you please confirm?

gmarull replied:
The way to _hide_ content from Doxygen:

```c
/** @cond INTERNAL_HIDDEN */

/* whatever here */

/** @endcond */
```

carlescufi replied:
But we don't really want to hide it This is not in `include/` so Doxygen will not process it. What we want is to mark it as an internal API (documented with Doxygen) so that if, in the future, we decide to start compiling internal APIs with Doxygen, it'll be ready @gmarull.

gmarull replied:
So using `@internal`/`@endinternal` allows making certain fragments internal, but afaik it will not hide the entity being documented.

carlescufi replied:
@ahmedmoheb-nordic why did you set `@code INTERNAL_DOCS`? That´s not what we proposed is it?

ahmedmoheb-nordic replied:
> @ahmedmoheb-nordic why did you set `@code INTERNAL_DOCS`? That´s not what we proposed is it?

Sorry, I got confused.

So, I should wrap the comment with these `@internal/@endinternal`, correct?

ahmedmoheb-nordic replied:
@carlescufi I searched for similar uses and I found that only `@internal` needs to added to the comment section that you don't want to handle, please confirm?

gmarull replied:
My understanding here is that you want to exclude a set of APIs from Doxygen, am I right? Then just use https://github.com/zephyrproject-rtos/zephyr/pull/49596/files#r961716485. You can put multiple functions inside a cond/endcond block.

ahmedmoheb-nordic replied:
> @ahmedmoheb-nordic why did you set `@code INTERNAL_DOCS`? That´s not what we proposed is it?

I've replaced it with `@internal` which I think it has the same effect as wrapping the comment section with `/** @cond INTERNAL_DOCS */` `/** @endcond */` except it's simpler.

carlescufi replied:
Turns out this doesn't work @ahmedmoheb-nordic, sorry about that, my mistake. Can you please instead use `/** @cond INTERNAL_HIDDEN */` and `/** @endcond */`, but only once in the file, like here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nordic_nrf/common/pinctrl_soc.h#L23

ahmedmoheb-nordic committed:
bluetooth: host: keys: Rename bt_keys_foreach() to bt_keys_foreach_type()

Use the name bt_keys_foreach_type() to be more descriptive.

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>
